pkg/bitlpm/trie.go

```diff
@@ -0,0 +1,536 @@
+	// the prefix-key argument pair. If the Ancestors function argument
+	// returns false the iteration will stop. Ancestors will iterate
+	// keys from shortest to longest prefix match (that is, the
+	// longest match will be returned last).
+	//
+	// Note: If the prefix argument exceeds the Trie's maximum
+	// prefix, it will be set to the Trie's maximum prefix.
+	Ancestors(prefix uint, key K, fn func(uint, K, T) bool)
+	// Descendants iterates over every prefix-key pair that is contained
+	// by the prefix-key argument pair. If the Descendants function argument
+	// returns false the iteration will stop. Descendants does **not** iterate
+	// over matches in any guaranteed order.
+	//
+	// Note: If the prefix argument exceeds the Trie's maximum
+	// prefix, it will be set to the Trie's maximum prefix.
+	Descendants(prefix uint, key K, fn func(uint, K, T) bool)
+	// Upsert updates or inserts the trie with a a prefix, key,
+	// and value.
+	//
+	// Note: If the prefix argument exceeds the Trie's maximum
+	// prefix, it will be set to the Trie's maximum prefix.
+	Upsert(prefix uint, key K, value T)
+	// Delete removes a key with the exact given prefix and returns
+	// false if the key was not found.
+	//
+	// Note: If the prefix argument exceeds the Trie's maximum
+	// prefix, it will be set to the Trie's maximum prefix.
+	Delete(prefix uint, key K) bool
+	// Len returns the number of entries in the Trie
+	Len() uint
+	// ForEach iterates over every element of the Trie.
+	// If the function argument returns false the iteration
+	// stops.
+	ForEach(fn func(uint, K, T) bool)
+}
+
+// Key is an interface that implements all the necessary
+// methods to index and retrieve keys.
+type Key[K any] interface {
+	// CommonPrefix returns the amount of bits that
+	// are the same between this key and the argument
+	// value, starting from MSB.
+	CommonPrefix(K) uint
+	// BitValueAt returns the value of the bit at an argument
+	// index. MSB is 0 and LSB is n.
+	BitValueAt(uint) uint8
+	// Value returns the underlying value of the Key.
+	Value() K
+}
+
+// trie is the generic implementation of a bit-trie that can
+// accept arbitrary keys conforming to the Key[K] interface.
+type trie[K, T any] struct {
+	root      *node[K, T]
+	maxPrefix uint
+	entries   uint
+}
+
+// NewTrie returns a Trie that accepts the Key[K any] interface
+// as its key argument. This enables the user of this Trie to
+// define their own bit-key.
+func NewTrie[K, T any](maxPrefix uint) Trie[Key[K], T] {
+	return &trie[K, T]{
+		maxPrefix: maxPrefix,
+	}
+}
+
+// node represents a specific key and prefix in the trie
+type node[K, T any] struct {
+	children     [2]*node[K, T]
+	prefixLen    uint
+	key          Key[K]
+	intermediate bool
+	value        T
+}
+
+// Lookup returns the longest prefix match to a specific
+// key.
+func (t *trie[K, T]) Lookup(k Key[K]) T {
+	// default return value
+	var ret T
+	t.Ancestors(t.maxPrefix, k, func(_ uint, _ Key[K], v T) bool {
```

Comment:
sidenote: I idly wondered whether for specific use cases, it may be more optimal to do a longest prefix match starting from the current prefix length and going up to the root rather than starting from the root and iterating deeper to the target lookup key. That said, I think this is probably too deep of micro-optimizations to bother with for an initial implementation. Probably depends a lot on how populated the tree is, key lengths, and a bunch of other factors.

Reply:
There's a whitepaper on this that I'm working through right now. In theory, we should be able to implement it without modifying the interface of the data structure. It also eliminates the need for intermediate nodes.